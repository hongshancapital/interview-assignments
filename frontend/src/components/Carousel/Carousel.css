html{
  font-size: 100px;
}
.wrapper {
  height: 100vh;
  overflow: hidden;
  position: relative;
}
.imgList {
  display: flex;
  width: 100%;
  height: 100%;
  transition: all 0.8s ease;
}
.imgItem {
  width: 100%;
  flex: none;
  height: 100%;
  background-size: auto 100%;
  background-position-y: bottom;
  background-position-x: center;
  background-repeat: no-repeat;
}
.imgTitle {
  font-size: 0.6rem;
  padding: 1rem 0.3rem 0.2rem 0.3rem;
  margin: 0 auto;
}
.imgDes {
  font-size: 0.3rem;
  padding: 0rem 0.3rem 0 0.3rem;
  margin: 0 auto;
}
.dotList {
  position: fixed;
  bottom: 10px;
  left: 50%;
  display: flex;
  transform: translate(-50%, -50%);
}
.dotItem {
  width: 70px;
  height: 3px;
  background-color: #a9a9a9;
  border-radius: 3px;
  margin: 5px;
  cursor: pointer;
}
.inner {
  height: 100%;
  border-radius: 3px;
}
@keyframes progress {
  from {
    width: 0px;
  }
  to {
    width: 70px;
  }
}
.line {
  width: 100%;
  background-color: #ffffff;
}
