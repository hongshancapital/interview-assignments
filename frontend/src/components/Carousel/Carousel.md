## API

| 参数            | 说明                                             | 类型                      | 默认值 |
| --------------- | ------------------------------------------------ | ------------------------- | ------ |
| autoplay        | 是否自动切换，值为数字时是轮播间隔时间 单位是 ms | boolean                   | false  |
| speed           | 切花轮播时的过渡动画时长 单位是 ms               | number                    | 300    |
| showSwitch      | 是否显示左右切换按钮                             | boolean                   | false  |
| dots            | 是否显示面板指示点                               | boolean                   | true   |
| className       | 轮播组件根元素样式名称                           | string                    |        |
| onChangeCurrent | 切换面板的回调                                   | (current: number) => void |        |

## 方法

| 名称   | 描述           |
| ------ | -------------- |
| next() | 切换到下一面板 |
| prev() | 切换到上一面板 |
