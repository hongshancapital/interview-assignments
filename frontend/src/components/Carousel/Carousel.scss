.carousel-viewport {
    overflow: hidden;
    width: 100%;
    height: 100%;
    position: relative;
    .carousel-wrapper{
        display: inline-block;
        white-space: nowrap;
        height: 100%;
        width: 100%;
        .carousel-item{
            display: inline-block;
            flex: 1;
            width: 100vw;
            height: 100%;
         }
    }
    .carousel-indicator-wrapper{
        position: absolute;
        width: 100%;
        height: 3px;
        bottom: 5%;
        &.hide{
            display: none;
        }
    }
}
