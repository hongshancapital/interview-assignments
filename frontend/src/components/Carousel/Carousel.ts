
import EventEmitter from "../../utils/EventEmitter";
import {Options} from "./Options";

export default class Carousel extends EventEmitter {
  // 配置项
  private readonly options: Options;

  // 容器
  private readonly el: HTMLElement;
  private slideWrapEl: HTMLElement;
  
  private pagenationWrapEl: HTMLElement;
  private slides: NodeListOf<Element>;
  
  // 容器宽度
  private readonly elWidth: number;

  private activeIndex = 0;

  // 是否正在过度
  private isTransiting = false;
  
  private timer: any;

  constructor(el: HTMLElement | string, options?: Partial<Options>) {
    super(['transitionEnd']);
    if (!el) return;
    this.el = (typeof el === 'string' ? document.querySelector(el) : el) as HTMLElement;
    this.elWidth = this.el.offsetWidth;
    this.slideWrapEl = this.el.querySelector('.carousel-wrap') as HTMLElement;

    // 合并选项
    this.options = new Options().merge(options);
    this.init();
  }

  private init() {
    this.initPagenation();
    this.el.appendChild(this.pagenationWrapEl);
    // 初始化事件
    this._initEvents();
    this.autoPlay();
  }

  // 初始化分页器
  private initPagenation() {
    this.slides = this.slideWrapEl.querySelectorAll('.carousel-slide');
    const dotNum = this.slides.length;
    let wrapLen = dotNum;

    if (this.options.loop) {
      // 复制首尾两个slide分别放在slideWrapEl的尾首
      const firstNodeCopy = this?.slideWrapEl?.firstElementChild?.cloneNode(true);
      const lastNodeCopy = this?.slideWrapEl?.lastElementChild?.cloneNode(true);
      if(lastNodeCopy) {
        this.slideWrapEl.insertBefore(lastNodeCopy, this.slideWrapEl.firstElementChild);
      }

      if(firstNodeCopy) {
        this.slideWrapEl.appendChild(firstNodeCopy);
      }
      wrapLen += 2;
      this.slideWrapEl.style.left = -this.elWidth + 'px';
    }

    this.slideWrapEl.style.width = wrapLen * 100 + '%';

    // 分页器
    this.pagenationWrapEl = document.createElement('div');
    this.pagenationWrapEl.className = 'carousel-pagination-wrap';
    
    let iEl:HTMLElement;

    for (let a = 0; a < dotNum; a++) {
      iEl = document.createElement('div');
      iEl.className = (a === this.activeIndex ? 'progress-active progress' : 'progress');
      this.pagenationWrapEl.appendChild(iEl);
    }
  }


  // 绑定事件
  private _initEvents() {
    const pagenations = this.pagenationWrapEl.children;
   
    for(let a = 0; a < pagenations.length; a++) {
      pagenations[a].addEventListener('click', () => this.navigation(a));
    }

    this.slideWrapEl.addEventListener('webkitTransitionEnd', this.onTransitionEnd.bind(this));
    // this.el.addEventListener('mouseenter', this.clearInterval.bind(this));
    this.el.addEventListener('mouseleave', this.autoPlay.bind(this));
    this.emitEvent('onInit');
  }

  private go(dir: number) {
    if (this.isTransiting) return;
    this.isTransiting = true;
    this.slideWrapEl.style.transitionDuration = this.options.speed + 's';
    const len = this.slides.length;
    const index = dir > 0 ? this.activeIndex + 1 : this.activeIndex - 1;
    this.activeIndex = (index + len) % len;
      if (this.options.loop) {
        this.loop(dir);
      }else{
        this.slideWrapEl.style.left = -this.activeIndex * this.elWidth + 'px';
      }
    this.updateDots(this.activeIndex);
  }
  

  private loop(dir: number) {
    const oldLeft = this.getLeft(this.slideWrapEl);
    
    // 这里可以利用下dir的正负
    // const newLeft = dir > 0 ? oldLeft - this.elWidth : oldLeft + this.elWidth;
    const newLeft = oldLeft + this.elWidth * -dir;
    this.slideWrapEl.style.left = newLeft + 'px';
  }

  // 点击圆点导航
  private navigation(index: number) {
    if (this.activeIndex !== index) {
      this.slideWrapEl.style.transitionDuration = this.options.speed + 's';
      this.activeIndex = index;
      this.slideWrapEl.style.left = -(this.activeIndex + 1) * this.elWidth + 'px';
      this.updateDots(index);
    }
  }

  // 底部圆点
  private updateDots(index: number) {
    // console.log(index);
    const dots = this.pagenationWrapEl.childNodes;
    for (let a = 0; a < dots.length; a++) {
      (dots[a] as HTMLElement).classList.remove('progress-active');
      while (dots[a].firstChild) {
        dots[a].removeChild(dots[a].firstChild as HTMLElement);
      }
    }
    (dots[index] as HTMLElement).classList.add('progress-active');
    
    let progressChild:HTMLElement;
    progressChild = document.createElement('div');
    progressChild.className="progress-child";
    dots[index].appendChild(progressChild);
  }

  private onTransitionEnd() {
    // console.log('end');
    if (this.options.loop) {
      const left = this.getLeft(this.slideWrapEl);
      if (left >= 0) {
        this.slideWrapEl.style.transitionDuration = '0s';
        this.slideWrapEl.style.left = '-4000px';
        // console.log('跳最后一个');
      }else if (left <= -(this.slides.length + 1) * this.elWidth) {
        this.slideWrapEl.style.transitionDuration = '0s';
        this.slideWrapEl.style.left = '-800px';
        // console.log('跳第一个');
      }
    }
    this.isTransiting = false;
    
    // 发射选中事件，响应DatePicker.on方法
    this.trigger('transitionEnd', this.activeIndex);
    
    this.emitEvent('onTransitionEnd', this.activeIndex);
  }
  
  private autoPlay() {
    if (this.options.autoplay) {
      this.timer = setInterval(() => this.go(1), this.options.delay);
    }
  }
  
  private clearInterval() {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  private getLeft(el: HTMLElement): number{
    return Number(el.style.left.slice(0, -2));
  }

  // 发射自定义事件
  private emitEvent(type: string, ...args: any[]) {
    if (this.options[type]) {
      this.options[type](args);
    }
  }

}