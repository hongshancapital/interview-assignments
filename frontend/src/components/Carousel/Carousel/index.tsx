import React, { useEffect, useMemo, useState } from 'react';
import { CarouselItem, CarouselItemProps } from '../CarouselItem';
import { Dots } from '../Dots';
import style from './index.module.scss';

export interface CarouselProps {
  interval?: number;
  showDots?: boolean;
  data: CarouselItemProps[];
}

export function Carousel(props: CarouselProps) {
  const { showDots, data, interval = 5000 } = props;

  const [current, setCurrent] = useState<number>(-1);

  // 组件挂载之后再设置 current
  useEffect(() => {
    if (data.length) {
      setCurrent(1);
    }
  }, [data]);

  useEffect(() => {
    const timer = setTimeout(() => {
      const now = current >= data.length ? 1 : current + 1;
      setCurrent(now);
    }, interval);

    return () => clearTimeout(timer);
  }, [current, data, interval]);

  const width = useMemo(() => `${data.length * 100}%`, [data]);

  const left = useMemo(() => `-${(current - 1) * 100}%`, [current]);

  return (
    <div className={style['my-carousel']}>
      <div
        className={style.list}
        style={{ width, left, transition: `300ms linear` }}
      >
        {data.map((item) => (
          <CarouselItem {...item} key={item.id} />
        ))}
      </div>

      {showDots && data.length > 1 && (
        <Dots
          setCurrent={setCurrent}
          current={current}
          interval={interval}
          count={data.length}
          data={data}
        />
      )}
    </div>
  );
}
