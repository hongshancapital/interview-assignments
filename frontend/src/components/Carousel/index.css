html,body,#root{
    width: 100%;
    height: 100%;
}
.Carousel{
    width: 1000px;
    height: 600px;
    position: absolute;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    margin: auto;
    overflow: hidden;
}

.box{
    display:flex;
    position: absolute;
    left: 0;
    top: 0;
    transition:left .5s;
   
}

.content{
    width: 100%;
    height: 100%;
    position: relative;
}
img{
    width: 1000px;
    height: 600px;
}
.pointBox{
    position: absolute;
    left: 0;
    right: 0;
    bottom: 20px;
    display: flex;
    justify-content: center;
    align-items: center;
   
}
.point{
    width: 80px;
    height: 3px;
    background-color: rgb(114, 112, 112);
    margin-right: 3px;
}
.pointInner{
    background-color: rgb(223, 213, 213);
    width:0;
    height: 100%;
    
}
.title{
    position: absolute;
    left: 0;
    right: 0;
    top: 170px;
    color: #fff;
    font-size: 16px;
}
.titleOther{
    color: #000;
}
.titleName{
    font-size: 50px;
}
