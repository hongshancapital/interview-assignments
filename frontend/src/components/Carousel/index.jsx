import React, { useState, useEffect } from "react";
import { useObserver } from "mobx-react";

import "./style.css";

const dotPositions = ["top", "bottom", "left", "right"];

export default function Carousel(props) {
  const {
    imgList = [],
    autoplay = true, //是否自动切换(默认true)
  } = props;

  //轮播列表索引(默认为0，第一个开始)
  const [activeNumber, setActiveNumber] = useState(0);

  //轮播时间间隔，默认为2秒
  const [times, setTimes] = useState(2000);

  const imgLen = useState(imgList && imgList.length ? imgList.length : 0);
  //切换到指定面板
  const goTo = (slideNumber) => {
    if (slideNumber < 0) {
      setActiveNumber(0);
    } else if (imgList && slideNumber > imgList.length - 1) {
      setActiveNumber(0);
    } else {
      setActiveNumber(slideNumber);
    }
  };

  //切换到下一面板
  const next = () => {
    setActiveNumber(activeNumber + 1);
  };

  //切换到上一面板
  const prev = () => {
    setActiveNumber(activeNumber - 1);
  };

  useEffect(() => {
    if (autoplay) {
      const interval = setInterval(() => {
        goTo(activeNumber + 1);
      }, times);
      return () => {
        if (interval) {
          clearInterval(interval);
        }
      };
    }
  });

  //加载图片和描述内容
  const imgLisCom = () => {
    if (imgList && imgList.length > 0) {
      return imgList.map((item, index) => {
        const {
          titile,
          describe,
          fit = "cover",
          width = "100%",
          height = "100%",
        } = item;
        
        return (
          <div
            className={`slick-slide ${
              activeNumber === index && "slick-active"
            }`}
            key={index}
          >
            <div className="slick-text">
              <div className="slick-title" style={{ color: item.fontColor }}>
                {titile}
              </div>
              <div className="slick-describe" style={{ color: item.fontColor }}>
                {describe}
              </div>
            </div>

            <img
              src={item.src}
              style={{
                objectFit: fit,
                width: width,
                height: height,
              }}
              alt={titile}
            ></img>
          </div>
        );
      });
    } else {
      return <></>;
    }
  };

  //加载底部dots
  const slickDotsCom = () => {
    if (imgList && imgList.length > 0) {
      return imgList.map((item, index) => {
        const { butBackColor = '' } = item;

        return (
          <li
            className={`${
              activeNumber === index && "slick-active"
            }`}
            style={{ backgroundColor: butBackColor }}
            key={index}
          >
            <button onClick={() => goTo(index)}>{index}</button>
          </li>
        );
      });
    } else {
      return <></>;
    }
  };

  return useObserver(() => (
    <div className="my-carousel">
      <div className="slick-slider" style={{ backgroundColor: imgList[activeNumber]?.backColor }} >
        <div className="slick-list">{ imgLisCom() }</div>

        <ul className={`slick-dots slick-dots-bottom`} style={{ width: `{ ${imgLen * 36}}px` }} >
          {slickDotsCom()}
        </ul>
      </div>
    </div>
  ));
}
