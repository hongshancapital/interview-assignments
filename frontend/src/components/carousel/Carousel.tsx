import CarouselWrap from './CarouselWrap'
import React, { useMemo } from 'react'
import { CarouselConfigs } from './types'
interface CarouselProps extends CarouselConfigs {
  children: ValidReactChild | ValidReactChild[]
}

const Carousel: React.FC<CarouselProps> = ({ children, ...rest }) => {
  const validChildren = useMemo(() => {
    const childrenOfArray = React.Children.toArray(children)
    return childrenOfArray.filter((child) => {
      if (typeof child === 'string') {
        return !!child.trim()
      }
      return !!child
    })
  }, [children])

  return (
    <>
      <CarouselWrap {...rest}>{validChildren}</CarouselWrap>
    </>
  )
}

export default Carousel
