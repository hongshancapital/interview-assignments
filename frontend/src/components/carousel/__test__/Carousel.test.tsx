import renderer from 'react-test-renderer';
import Carousel from '../Carousel';
import { render } from '@testing-library/react';

describe('Carousel', () => {

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('component could be updated and unmounted without errors', () => {
        const { unmount, rerender } = render(<Carousel><h1>hello</h1><h1>world</h1></Carousel>);
        expect(() => {
            rerender(<Carousel><h1>hello</h1><h1>world</h1></Carousel>);
            unmount();
        }).not.toThrow();
    });

    it('snapshot', () => {
        const component = renderer.create(
            <Carousel><h1>hello</h1><h1>world</h1></Carousel>,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})

