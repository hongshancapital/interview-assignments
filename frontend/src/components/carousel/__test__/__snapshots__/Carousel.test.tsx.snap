// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel snapshot 1`] = `
<div
  className="carousel-wrap"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="carousel-track"
    style={
      Object {
        "transform": "translate3d(-000%, 0, 0)",
      }
    }
  >
    <div
      className="carousel-slide"
      data-index={0}
      style={Object {}}
      tabIndex="-1"
    >
      <h1>
        hello
      </h1>
    </div>
    <div
      className="carousel-slide"
      data-index={1}
      style={Object {}}
      tabIndex="-1"
    >
      <h1>
        world
      </h1>
    </div>
  </div>
  <div
    className="carousel-dots"
  >
    <div
      className="carousel-dot"
      onClick={[Function]}
    >
      <div
        className="carousel-dot-inner"
      >
        <div
          className="carousel-dot__process"
          style={Object {}}
        />
      </div>
    </div>
    <div
      className="carousel-dot"
      onClick={[Function]}
    >
      <div
        className="carousel-dot-inner"
      >
        <div
          className="carousel-dot__process"
          style={Object {}}
        />
      </div>
    </div>
  </div>
</div>
`;
